components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java

```diff
@@ -697,18 +697,22 @@ private String handleUserConsent(HttpServletRequest request, String consent, OAu
             return EndpointUtil.getErrorRedirectURL(oauthProblemException, oauth2Params);
         }
 
-        //When responseType equal to "id_token" the resulting token is passed back as a query parameter
-        //According to the specification it should pass as URL Fragment
-        if (OAuthConstants.ID_TOKEN.equalsIgnoreCase(responseType)) {
-            if (authzRespDTO.getCallbackURI().contains("?")) {
-                return authzRespDTO.getCallbackURI() + "#" + oauthResponse.getLocationUri().substring(
-                        authzRespDTO.getCallbackURI().length() + 1, oauthResponse.getLocationUri().length());
+        //When response_mode equals to form_post, body parameter is passed back.
+        if (RESPONSE_MODE_FORM_POST.equals(oauth2Params.getResponseMode()) && oauthResponse.getBody() != null) {
+            return oauthResponse.getBody();
+        } else {
+            //When responseType equal to "id_token" the resulting token is passed back as a query parameter
+            //According to the specification it should pass as URL Fragment
+            if (OAuthConstants.ID_TOKEN.equalsIgnoreCase(responseType)) {
+                if (authzRespDTO.getCallbackURI().contains("?")) {
+                    return authzRespDTO.getCallbackURI() + "#" + oauthResponse.getLocationUri().substring(
+                            authzRespDTO.getCallbackURI().length() + 1, oauthResponse.getLocationUri().length());
```

Comment:
oauthResponse.getLocationUri().substring(authzRespDTO.getCallbackURI().length() + 1) should be sufficient right?

better to use StringUtils.substring(oauthResponse.getLocationUri(), authzRespDTO.getCallbackURI().length() + 1)